Bulldozer.CSV/CSVComponent.cs

```diff
@@ -488,804 +836,2680 @@ public static int GetAttributeFieldType( string attributeTypeString )
+                                                                                .ToDictionary( k => $"{k.EntityTypeId}_{k.Key}", v => definedTypeDict.GetValueOrNull( v.AttributeQualifiers.FirstOrDefault( aq => aq.Key == "definedtype" ).Value.AsIntegerOrNull().Value ) );
+            
+            var attributeDefinedValuesDict = attributeDefinedTypeDict.ToDictionary( k => k.Key, v => v.Value.DefinedValues.ToDictionary( a => a.Value, b => b.DefinedTypeId ) );
+
+            var attributeValuesToProcess = attributeValues.Where( v => attributeDefinedValuesDict.ContainsKey( $"{v.EntityTypeId}_{v.AttributeKey}" ) && !attributeDefinedValuesDict[$"{v.EntityTypeId}_{v.AttributeKey}"].ContainsKey( v.AttributeValue  ) )
+                                                          .Select( v => new AttributeValueObject { AttributeKey = v.AttributeKey, AttributeValue = v.AttributeValue, EntityTypeId = v.EntityTypeId, DefinedTypeId = attributeDefinedTypeDict.GetValueOrNull( $"{v.EntityTypeId}_{v.AttributeKey}" )?.Id } )
+                                                          .GroupBy( o => new { o.DefinedTypeId, o.AttributeValue } )
+                                                          .Select( grp => grp.First() )
+                                                          .OrderBy( g => g.DefinedTypeId )
+                                                          .ToList();
+
+            var definedValuesToAdd = new List<DefinedValue>();
+            var attributeKey = string.Empty;
+            DefinedTypeCache attributeDefinedType = null;
+
+            if ( attributeValuesToProcess.Count > 0 )
+            {
+                ReportProgress( 0, $"Creating {attributeValuesToProcess.Count} new DefinedValues..." );
 
-        #region Group Constants
+                foreach ( var attributeValue in attributeValuesToProcess )
+                {
+                    if ( attributeValue.AttributeKey != attributeKey )
+                    {
+                        attributeDefinedType = attributeDefinedTypeDict.GetValueOrNull( $"{attributeValue.EntityTypeId}_{attributeValue.AttributeKey}" );
+                    }
 
-        /*
-         * This is the definition of the csv format for the Group.csv file
-         */
+                    var attributeDefinedValue = attributeDefinedType?.DefinedValues.FirstOrDefault( dv => dv.Value == attributeValue.AttributeValue );
+                    if ( attributeDefinedValue != null )
+                    {
+                        continue;
+                    }
+                    else
+                    {
+                        var newDefinedValue = new DefinedValue
+                        {
+                            DefinedTypeId = attributeDefinedType.Id,
+                            Value = attributeValue.AttributeValue,
+                            Order = 0,
+                            ForeignKey = this.ImportInstanceFKPrefix
```

Comment:
Should this FK be more complex?

Reply:
No context for a source driven foreign key. At least capturing the prefix marks it.